src/compute-client/src/controller/sequential_hydration.rs

```diff
@@ -0,0 +1,262 @@
+//! controller and the `PartitionedState` client that splits commands across replica processes.
+//! This location is important:
+//!
+//!  * It needs to be behind the controller since hydration is a per-replica mechanism. Different
+//!    replicas can progress through hydration at different paces.
+//!  * It needs to be before the `PartitionedState` client because all replica workers must see
+//!    `Schedule` commands in the same order. Otherwise we risk getting stuck when different
+//!    workers hydrate different dataflows and wait on each other for progress in these dataflows.
+//!  * It also needs to be before the `PartitionedState` client because it needs to be able to
+//!    observe all compute commands. Clients behind `PartitionedState` are not guaranteed to do so,
+//!    since commands are only forwarded to the first process.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use mz_compute_types::dyncfgs::HYDRATION_CONCURRENCY;
+use mz_dyncfg::ConfigSet;
+use mz_ore::collections::CollectionExt;
+use mz_ore::soft_assert_eq_or_log;
+use mz_repr::GlobalId;
+use mz_service::client::GenericClient;
+use timely::progress::{Antichain, Timestamp};
+use timely::PartialOrder;
+use tracing::debug;
+
+use crate::protocol::command::ComputeCommand;
+use crate::protocol::response::ComputeResponse;
+use crate::service::ComputeClient;
+
+/// A shareable token.
+type Token = Arc<()>;
+
+/// A client enforcing sequential dataflow hydration.
+#[derive(Debug)]
+pub(super) struct SequentialHydration<C, T> {
+    /// The wrapped client.
+    client: C,
+    /// Dynamic system configuration.
+    dyncfg: Arc<ConfigSet>,
+    /// Tracked collections.
+    ///
+    /// Entries are inserted in response to observed `CreateDataflow` commands.
+    /// Entries are removed in response to `FrontierUpper` commands that report collection
+    /// hydration, or in response to `AllowCompaction` commands that specify the empty frontier.
+    collections: BTreeMap<GlobalId, Collection<T>>,
+    /// A queue of scheduled collections that are awaiting hydration.
+    hydration_queue: VecDeque<GlobalId>,
+    /// A token held by hydrating collections.
+    ///
+    /// Useful to efficiently determine how many collections are currently in the process of
+    /// hydration, and thus how much capacity is available.
+    hydration_token: Token,
+}
+
+impl<C, T> SequentialHydration<C, T>
+where
+    C: ComputeClient<T>,
+    T: Timestamp,
+{
+    /// Create a new `SequentialHydration` client.
+    pub fn new(client: C, dyncfg: Arc<ConfigSet>) -> Self {
+        Self {
+            client,
+            dyncfg,
+            collections: Default::default(),
+            hydration_queue: Default::default(),
+            hydration_token: Default::default(),
+        }
+    }
+
+    /// Return the number of hydrating collections.
+    fn hydration_count(&self) -> usize {
+        Arc::strong_count(&self.hydration_token) - 1
```

Comment:
Is this equivalent to:
```
self.collections.values().filter(|c| matches!(c.state, State::Hydrating(_))).count()
```
? If yes, although slower at runtime, maybe it's more obvious than using an arc?

Reply:
Yes, it is. And I agree it would be more obvious, but I'm a bit worried about all that iterating. `hydration_count` is invoked every time `hydrate_collections` is invoked, which happens every time we receive a `ComputeCommand`. Of course we can write the code to only count hydrating collections if necessary, but it does complicate the logic in other places.

How much do you dislike the `Arc`? If the answer is "not a lot" I'd rather keep it, but happy to change things otherwise!

Reply:
Nope, `Arc` is fine with me -- it was me trying to figure out whether I was missing something. Thanks for the explanation.